numpy/_core/src/common/raii_utils.hpp

```diff
@@ -0,0 +1,151 @@
+#ifndef NUMPY_CORE_SRC_COMMON_GIL_AND_THREAD_HPP_
+#define NUMPY_CORE_SRC_COMMON_GIL_AND_THREAD_HPP_
+
+//
+// Utiltities for RAII management of resources.
+//
+// Another (and arguably clearer) name for this resource management pattern
+// is "Scope-Bound Resource Management", but RAII is much more common, so we
+// use the familiar acronym.
+//
+
+#include <Python.h>
+
+// For npy_string_allocator, PyArray_StringDTypeObject, NPY_NO_EXPORT:
+#include "numpy/ndarraytypes.h"
+
+// Forward declarations not currently in a header.
+// XXX Where should these be moved?
+NPY_NO_EXPORT npy_string_allocator *
+NpyString_acquire_allocator(const PyArray_StringDTypeObject *descr);
+NPY_NO_EXPORT void
+NpyString_release_allocator(npy_string_allocator *allocator);
+
+
+namespace np { namespace raii {
+
+//
+// RAII for PyGILState_* API.
+//
+// In C++ code, use this at the beginning of a scope, e.g.
+//
+//     {
+//         np::raii::EnsureGIL ensure_gil{};
+//         [code that uses the Python C API here]
+//     }
+//
+// instead of
+//
+//     PyGILState_STATE gil_state =  PyGILState_Ensure();
+//     [code that uses the Python C API here]
+//     PyGILState_Release(gil_state);
+//
+// or
+//     NPY_ALLOW_C_API_DEF
+//     NPY_ALLOW_C_API
+//     [code that uses the Python C API here]
+//     NPY_DISABLE_C_API
+//
+// This ensures that PyGILState_Release(gil_state) is called,  even if the
+// wrapped code throws an exception or executes a return or a goto.
+//
+class EnsureGIL
+{
+    PyGILState_STATE gil_state;
+
+public:
+
+    EnsureGIL() {
+        gil_state = PyGILState_Ensure();
+    }
+
```

Comment:
You probably want `EnsureGIL(EnsureGIL const&) = delete;` too (and the assignment version, and perhaps also the `&&` version), to avoid doubly-releasing.

Reply:
Thanks, done.